src/tron/tronEngine.ts

```diff
@@ -0,0 +1,1060 @@
+// @ts-expect-error
+import { byteArray2hexStr } from '@tronscan/client/src/utils/bytes'
+// @ts-expect-error
+import { contractJsonToProtobuf } from '@tronscan/client/src/utils/tronWeb'
+import { add, div, eq, gt, lt, lte, mul, sub } from 'biggystring'
+import { asMaybe, Cleaner } from 'cleaners'
+import {
+  EdgeCurrencyEngineOptions,
+  EdgeFetchFunction,
+  EdgeLog,
+  EdgeSpendInfo,
+  EdgeTransaction,
+  EdgeWalletInfo,
+  InsufficientFundsError,
+  JsonObject,
+  NoAmountSpecifiedError
+} from 'edge-core-js/types'
+
+import { CurrencyEngine } from '../common/engine'
+import {
+  asyncWaterfall,
+  getDenomInfo,
+  getOtherParams,
+  hexToDecimal,
+  makeMutex,
+  padHex,
+  shuffleArray
+} from '../common/utils'
+import { TronPlugin } from './tronPlugin'
+import {
+  asAccountResources,
+  asBroadcastResponse,
+  asChainParams,
+  asEstimateEnergy,
+  asTransaction,
+  asTRC20Balance,
+  asTRC20Transaction,
+  asTRC20TransactionInfo,
+  asTriggerSmartContract,
+  asTronBlockHeight,
+  asTronQuery,
+  asTRXBalance,
+  asTRXTransferContract,
+  ReferenceBlock,
+  TronAccountResources,
+  TronNetworkFees,
+  TronSettings,
+  TronTxParams,
+  TxQueryCache
+} from './tronTypes'
+import {
+  base58ToHexAddress,
+  encodeTRC20Transfer,
+  hexToBase58Address,
+  TronScan
+} from './tronUtils'
+
+const queryTxMutex = makeMutex()
+
+const ACCOUNT_POLL_MILLISECONDS = 20000
+const BLOCKCHAIN_POLL_MILLISECONDS = 20000
+const TRANSACTION_POLL_MILLISECONDS = 3000
+const NETWORKFEES_POLL_MILLISECONDS = 60 * 10 * 1000
+
+type TronFunction =
+  | 'trx_blockNumber'
+  | 'trx_broadcastTx'
+  | 'trx_chainParams'
+  | 'trx_estimateEnergy'
+  | 'trx_getAccountResource'
+  | 'trx_getBalance'
+  | 'trx_getTransactionInfo'
+  | 'trx_getTransactions'
+
+export class TronEngine extends CurrencyEngine<TronPlugin> {
+  fetchCors: EdgeFetchFunction
+  log: EdgeLog
+  readonly recentBlock: ReferenceBlock
+  accountResources: TronAccountResources
+  networkFees: TronNetworkFees
+  settings: TronSettings
+  accountExistsCache: { [address: string]: boolean }
+  energyEstimateCache: { [addressAndContract: string]: number }
+  tronscan: any
+
+  constructor(
+    currencyPlugin: TronPlugin,
+    walletInfo: EdgeWalletInfo,
+    opts: EdgeCurrencyEngineOptions,
+    fetchCors: EdgeFetchFunction
+  ) {
+    super(currencyPlugin, walletInfo, opts)
+    this.fetchCors = fetchCors
+    this.log = opts.log
+    this.settings = this.currencyInfo.defaultSettings
+      .otherSettings as TronSettings
```

Comment:
We need to stop jamming our network info into `defaultSettings`. Take a look at the `EosNetworkInfo` type, and how I pass it to the engine / tools constructors. We need to rename `TronSettings` to `TronNetworkInfo` and pass it into the engine alongside the currency info, instead of inside it.

Reply:
Instead of threading this through and rebasing I just added as an additional commit